src/Bicep.Core/Emit/SourceMapHelper.cs

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Bicep.Core.Extensions;
+using Bicep.Core.Parsing;
+using Bicep.Core.Syntax;
+using Bicep.Core.Text;
+using Bicep.Core.Workspaces;
+using Microsoft.WindowsAzure.ResourceStack.Common.Extensions;
+
+namespace Bicep.Core.Emit
+{
+    public static class SourceMapHelper
+    {
+        public static void AddMapping(
+            IDictionary<string, IDictionary<int, IList<(int start, int end)>>> rawSourceMap,
+            BicepFile bicepFile,
+            SyntaxBase bicepSyntax,
+            PositionTrackingJsonTextWriter jsonWriter,
+            int jsonStartPos)
+        {
+            if (rawSourceMap is null)
+            {
+                throw new ArgumentNullException(nameof(rawSourceMap));
+            }
+
+            var bicepFilePath = bicepFile.FileUri.AbsolutePath;
+            (int bicepLine, _) = TextCoordinateConverter.GetPosition(bicepFile.LineStarts, bicepSyntax.GetPosition());
+
+            // account for leading nodes (decorators)
+            if (bicepSyntax is StatementSyntax syntax)
+            {
+                bicepLine += syntax.LeadingNodes.Count(node => node is Token token && token.Type is TokenType.NewLine);
+            }
+
+            // increment start position if starting on a comma (happens when outputting successive items in objects and arrays)
+            if (jsonWriter.CommaPositions.Contains(jsonStartPos))
+            {
+                jsonStartPos++;
+            }
+
+            var jsonEndPos = jsonWriter.CurrentPos - 1;
+
+            rawSourceMap.AddMapping(
+                bicepFilePath,
+                bicepLine,
```

Comment:
What about storing `IPositionable` instead of just the line number in the raw source map, and performing the conversion from position -> line numbers at the end when you're outputting the JSON?

Reply:
OK cool this is along the lines of what I was thinking of doing before but I was getting tripped up trying to account for the leading nodes and losing that context. Your other comment's suggestions about the leading nodes is giving me some additional clarity so I can look into this again.